L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,450 @@
+L46: C-core: New TLS Credentials API
+----
+* Author(s): ZhenLian
+* Approver: markdroth, yashykt, jiangtaoli2016
+* Status: Draft
+* Implemented in:
+* Last updated: September 9, 2020
+
+## Abstract
+
+This proposal aims to provide the following security features in a new TLS credential API:
+1. credential reloading: reload transport credentials periodically from various sources, e.g. from credential files, or users' customized Provider implementations, without shutting down
+2. custom authorization check: optionally bypass the certificate check, bypass the hostname check on the client side, and perform an additional per-connection authorization callback at the end of authentication
+3. TLS version configuration: optionally configure the minimum and maximum of the TLS version that will be used in the handshake
+
+## Background
+
+The current TLS implementation in gRPC C core has some restrictions:
+1. lack of the credential reloading support. Once an application starts, it can never reload its identity key/cert-chain and root certificates without shutting down
+2. inflexibility in the peer verification. In a mutual TLS scenario, servers would always perform certificate verification, and clients would always perform certificate verification as well as default hostname check. No customized checks can be introduced.
+3. inability to choose the TLS version that will be used
+
+There have always been some ad hoc requests on GitHub Issue Page for supporting these features. The demands for a more flexible and configurable API have increased with the emerging needs in cloud platform.
+One common example is to support SPIFFE ID as the identity format, which requires us to disable the default hostname check on the client side.
```

Comment:
SPIFFE ID and hostname verification check are orthogonal to each other. Nothing prevents a certificate containing both SPIFFE ID as URI and hostname as dnsName SAN fields, in which case, hostname verification check can still be performed on the dnsName field.

Reply:
Yeah by saying "default hostname check", I meant the exact check between the hostname sent from client and the identity shown on the server's certificate. If using SPIFFE, the identity is a SPIFFE ID and doesn't match the hostname sent from client. I added some explanations. Does it look better now?

Reply:
It still does not sound correct. Notice that in the hostname verification check, a client verifies that the dnsName in the subjectAltName field (i.e., hostname) of the certificate matches the host portion of the URL used to make the request. In other words, it does not compare SPIFFE ID with URL, and does not matter if they are different.

Reply:
Ah I see what you mean. IIUC, in typical SPIFFE semantics, a physical server can have multiple workload ids, and a valid SPIFFE cert might not contain dnsName in the SAN field. If we do default hostname verification on a SPIFFE cert, the check will fail. Isn't that the original problem we are going to solve? I might misunderstand something.
How about rewording as "a valid certificate using SPIFFE ID doesn't necessarily contain the dnsName in the SAN field, which might cause default hostname verification to fail"?

Reply:
That sounds much better. More importantly, I think we need to highlight that we need an ability of skipping the hostname verification check if a client has a different way of validating the server's identity.

Reply:
SG. Revised! Please take a look, thanks!